src/wallet/rpcdump.cpp

```diff
@@ -832,13 +837,18 @@ UniValue z_importviewingkey(const UniValue& params, bool fHelp)
         throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Invalid viewing key");
     }
 
+    auto addrInfo = boost::apply_visitor(libzcash::AddressInfoFromViewingKey{}, viewingkey);
+    UniValue result(UniValue::VOBJ);
+    result.pushKV("type", addrInfo.first);
+    result.pushKV("address", EncodePaymentAddress(addrInfo.second));
+
     auto addResult = boost::apply_visitor(AddViewingKeyToWallet(pwalletMain), viewingkey);
     if (addResult == SpendingKeyExists) {
         throw JSONRPCError(
             RPC_WALLET_ERROR,
             "The wallet already contains the private key for this viewing key");
     } else if (addResult == KeyAlreadyExists && fIgnoreExistingKey) {
-        return NullUniValue;
+        return result;
     }
```

Comment:
The fact that this is an error means that the caller can't rely on calling `z_importviewingkey` to return the corresponding address. So if you wanted to express "add this viewing key (even if the spending key might already exist in the wallet), and return the corresponding address", there's no easy way to do that without catching the RPC error and falling back to... well, there's no RPC call to fall back to that *just* computes the address for an existing viewing key. So you'd have to fall back to external library code.

The fact that we have `whenkeyisnew` as a (default!) rescan option indicates that this API is supposed to be usable regardless of whether the key is new, and I don't think it makes sense to interpret this only as relating to the viewing key rather than also the underlying spending key.

So, I think that adding a viewing key for a spending key that already exists should be ignored (except for possibly rescanning). This will need a test, and the change should be documented in the help string and release notes.

```suggestion
    result.pushKV("havespendingkey", UniValue(addResult == SpendingKeyExists));

    if ((addResult == SpendingKeyExists || addResult == KeyAlreadyExists) && fIgnoreExistingKey) {
        return result;
    }
```

Reply:
For Sapling, viewing keys are handled the same whether they are standalone or have an associated spending key. So it should be fine to allow importing a viewing key that already has an associated spending key.

The problem is that this is an interface change for Sprout, where the underlying key store handles viewing keys differently to spending keys, and makes assumptions about what is present where. We can't relax this RPC interface without also ensuring that is addressed (and it's past 1am so I can't check right now what happens after this part of the diff).

Reply:
Not sure what to do here.

Reply:
Let's make a separate issue for this potential change, and leave it out of this PR.